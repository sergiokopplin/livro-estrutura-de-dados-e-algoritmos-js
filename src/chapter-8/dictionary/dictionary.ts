export function defaultToString(item: string) {
  return item;
}

export class ValuePair {
  key: string;
  value: string;

  constructor(key: string, value: string) {
    this.key = key;
    this.value = value;
  }

  toString() {
    return `[#${this.key}: ${this.value}]`;
  }
}

export class Dictionary {
  table: {
    [key: string]: any;
  };
  toStrFn: any;

  constructor(toStrFn = defaultToString) {
    this.table = {};
    this.toStrFn = toStrFn;
  }

  set(key: string, value: any): boolean {
    if (key != null && value != null) {
      const tableKey: string = this.toStrFn(key);
      this.table[tableKey] = new ValuePair(key, value);

      return true;
    }

    return false;
  }

  has(key: string): boolean {
    return this.table[this.toStrFn(key)] != null;
  }

  size(): number {
    return Object.keys(this.table).length;
  }

  keys(): string[] {
    return Object.keys(this.table).map(item => this.table[item].key);
  }

  values(): string[] {
    return Object.keys(this.table).map(item => this.table[item].value);
  }

  get(key: string): string {
    return this.table[this.toStrFn(key)].value;
  }

  keyValues(): {
    [key: string]: any;
  }[] {
    return Object.keys(this.table).map(item => this.table[item]);
  }

  remove(key: string): void {
    delete this.table[this.toStrFn(key)];
  }
}
