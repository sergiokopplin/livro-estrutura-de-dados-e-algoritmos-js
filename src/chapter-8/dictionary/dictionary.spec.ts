import { Dictionary } from './dictionary';

describe('Dictionary', () => {
  test('Should construct correctly', () => {
    const dictionary = new Dictionary();

    expect(dictionary).toMatchSnapshot();
  });

  test('Should work properly', () => {
    const dictionary = new Dictionary();

    dictionary.set('Gandalf', 'gandalf@email.com');
    dictionary.set('John', 'john@email.com');
    dictionary.set('Tyrion', 'tyrion@email.com');

    expect(dictionary.has('Galdalf')).toBe(true);
    expect(dictionary.size()).toBe(3);
    expect(dictionary.keys()).toEqual(['Galdalf', 'John', 'Tyrion']);
    expect(dictionary.values()).toEqual([
      'gandalf@email.com',
      'john@email.com',
      'tyrion@email.com',
    ]);
    expect(dictionary.get('Tyrion')).toBe('tyrion@email.com');
    expect(dictionary.keyValues()).toEqual([
      {
        key: 'Gandalf',
        value: 'gandalf@email.com',
      },
      {
        key: 'John',
        value: 'john@email.com',
      },
      {
        key: 'Tyrion',
        value: 'tyrion@email.com',
      },
    ]);

    dictionary.remove('John');

    expect(dictionary.has('John')).toBe(false);
    expect(dictionary.size()).toBe(2);
    expect(dictionary.keys()).toEqual(['Galdalf', 'Tyrion']);
    expect(dictionary.values()).toEqual(['gandalf@email.com', 'tyrion@email.com']);
    expect(dictionary.keyValues()).toEqual([
      {
        key: 'Gandalf',
        value: 'gandalf@email.com',
      },
      {
        key: 'Tyrion',
        value: 'tyrion@email.com',
      },
    ]);
  });
});
