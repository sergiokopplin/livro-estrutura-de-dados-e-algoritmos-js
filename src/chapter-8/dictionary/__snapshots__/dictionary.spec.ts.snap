// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary Should construct correctly 1`] = `
Dictionary {
  "items": Object {},
}
`;

exports[`Dictionary Should work properly 1`] = `
Dictionary {
  "items": Object {
    "1": 1,
  },
}
`;

exports[`Dictionary Should work properly 2`] = `
Dictionary {
  "items": Object {
    "1": 1,
    "2": 2,
  },
}
`;

exports[`Dictionary Should work properly 3`] = `
Dictionary {
  "items": Object {
    "2": 2,
  },
}
`;

exports[`Dictionary Should work properly 4`] = `
Dictionary {
  "items": Object {},
}
`;
